### ISSUE TITLE
Make sure that all files in `tests/reference-files` are actually used

### ISSUE THREAD

**Issue by isteinbrecher** [beamme-py/beamme#106]
As of right now, we don't have a way of checking that all files in `tests/reference-files` are actually used during testing. This sometimes results in old, not used files being there without someone noticing. Let's discuss if there is an easy way to check that all reference files are actually used during testing.

**Comment by davidrudlstorfer** [comment]
I gave this problem some thoughts today and also tested a small strategy - unfortunately I was not succesfull and I think with our current test approach there's no easy way to check that all reference files are used.

1. The optimal way would be to check if all reference files are used by running a pre-commit hook. In our case this is not easily doable because the reference file paths are generated at runtime (with the pytest fixture `get_corresponding_reference_file_path`)

2. Therefore, we definitely must check the usage of all files during the execution of pytest. Again, the optimal solution would be on a file system basis to check if all files were being read - but that also turns out to be not really doable. To solve this problem I simply tracked the files during the pytest run (simply store the path of the used reference file in `get_corresponding_reference_file_path` into a global variable) and compared it with all reference files at the end of the run (a `pytest_sessionfinish` function is executed at the end of each pytest run). And this approach now shows us why it is not easily feasible =>

    2.1 In our testing strategy we also rely on hard coded paths to reference files which are, therefore, not tracked with the function `get_corresponding_reference_file_path` => these files won't be checked

    2.2 A few reference files are only used in specific testing setups - i.e., CubitPy or 4C. This means we would have to track the used reference files across all Github Actions, somehow combine them into a global list in a separate Github workflow and then we are able to check if all files are used.

This means we would have to 1. rewrite all file loads to be handled through one central function, i.e., `get_corresponding_reference_file_path` 2. temporarily write all utilized reference files into a file and upload it as an artifact and 3. create a new Github action which combines all used reference files and checks that all files are used

Summing up, in my opinion it is currently not easily feasible to check if all reference files are used by the testing setup. I propose to just check all of them manually now and remove any old outdated ones. In the future we just have to keep track of changes within the PR's.

@isteinbrecher maybe you have a different opinion or I missed a simpler approach - curious about your opinion

**Comment by isteinbrecher** [comment]
Thanks for the explanations @davidrudlstorfer. I actually had a look at this a while back (unfortunately never pushed the [branch](https://github.com/isteinbrecher/meshpy/tree/check-reference-files) or commented here...). Actually, I think I ended up at similar points as you did.

I think we have two options:
1. Leave it as it is, since the current approach is working, the only drawback is that we might "miss" a reference file that can be deleted.
1. Use the approach you described with tracking the files returned by `get_corresponding_reference_file_path`. The two issues there are (as you already pointed out):
    - We don't always use `get_corresponding_reference_file_path` when accessing a reference file. We could adapt this, that every file we access from the reference files is actually retrieved from `get_corresponding_reference_file_path`. In my opinion that would actually make the code "cleaner". We can then track the reference files and compare at the end of the test session that every file in the directory gets accesses.
    - We don't have a single test that tests all tests (for now or maybe forever). I think uploading the list to the artefacts and then doing something with that list is overkill for the current issue. What we could do, is check all files in the 4C/ArborX testing run, and explicitly add the files used by the cubit tests (should only be a couple) to an "allowlist".

I actually like the idea of extending the use of `get_corresponding_reference_file_path` to further clean up the naming in the reference directory. I think we kind of have a path to a solution to this issue drawn out, the question is if we want to go this way.

**Comment by isteinbrecher** [comment]
With #232 all files in the `reference_file` folder should be accessed via the `get_corresponding_reference_file_path` function.

**Comment by davidrudlstorfer** [comment]
Very nice that all reference file handling is now running through one central function:)

As we've discussed in person the approach to simply upload the used files (as a dictionary) as an artefact sounds viable and doable. I'll have a look once the src structure is done.

**Comment by isteinbrecher** [comment]
Sounds great! I am already curious if we have "ghost" files (and if we really get all files through `get_corresponding_reference_file_path` 😅 )
